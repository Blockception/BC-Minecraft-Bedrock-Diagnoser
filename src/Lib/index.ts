/*	Auto generated	*/
export * as Diagnostics from "./Diagnostics/index";
export * as Types from "./Types/index";
export * as Utility from "./utility/index";
