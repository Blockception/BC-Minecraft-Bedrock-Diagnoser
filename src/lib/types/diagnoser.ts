import { Pack, PackType, TextDocument } from "bc-minecraft-bedrock-project";
import { DiagnoserContext, ManagedDiagnosticsBuilder } from "./diagnoser-context";
import { DiagnosticSeverity } from "./severity";
import { format_diagnose_path } from "../diagnostics/format";
import { MCIgnore } from "bc-minecraft-project";
import { minecraft_language_diagnose } from "../diagnostics/minecraft";
import { BehaviorPack } from "../diagnostics/behavior-pack";
import { ResourcePack } from "../diagnostics/resource-pack";
import { SkinPack } from "../diagnostics/skin-pack/skin-pack";
import { WorldPack } from "../diagnostics/world-pack/world-pack";
import { DocumentDiagnosticsBuilder } from "./diagnostics-builder";

import path from "path";

/**The object that is responsible for diagnosing minecraft bedrock files*/
export class Diagnoser<T extends TextDocument = TextDocument> {
  /**The context needed to perform diagnostics*/
  readonly context: DiagnoserContext<T>;

  /**Create a new instance of Diagnoser
   * @param context The context needed to perform diagnostics*/
  constructor(context: DiagnoserContext<T>) {
    this.context = context;
  }

  /** Process and diagnoses the given document
   * @param doc The textdocument to process or the uri to the document
   * @returns `true` or `false` if the diagnostics was successfully*/
  process(doc: T | string): boolean {
    if (typeof doc === "string") {
      const temp = this.context.getDocument(doc);

      if (!temp) return false;
      doc = temp;
    }

    const pack = this.context.getProjectData().projectData.get(doc);
    if (!pack) return false;

    //Check if diagnostics was disabled
    const ext = path.extname(doc.uri);

    //Check if diagnostics for this file type is disabled
    if (pack.context.attributes["diagnostic" + ext] === "false") return false;

    const diagnoser = this.context.getDiagnoser(doc, pack.context) as DocumentDiagnosticsBuilder<T> &
      ManagedDiagnosticsBuilder<T>;
    if (!diagnoser) return false;

    diagnoser.document = doc;
    let out = false;

    try {
      //diagnose path
      format_diagnose_path(pack, doc.uri, diagnoser);

      //Language file?
      if (doc.uri.endsWith(".lang")) {
        minecraft_language_diagnose(diagnoser, pack.type);
      } else {
        //Check per pack
        switch (pack.type) {
          case PackType.behavior_pack:
            out = BehaviorPack.diagnose_document(diagnoser);
            break;

          case PackType.resource_pack:
            out = ResourcePack.diagnose_document(diagnoser);
            break;

          case PackType.skin_pack:
            out = SkinPack.Process(diagnoser);
            break;

          case PackType.world:
            out = WorldPack.Process(diagnoser);
            break;
        }
      }
    } catch (err: any) {
      let msg;

      if (err.message && err.stack) {
        msg = `${err.message}\n\t${err.stack}`;
      } else {
        msg = JSON.stringify(err);
      }

      diagnoser.add({ character: 0, line: 0 }, msg, DiagnosticSeverity.error, "debugger.internal.exception");
    }

    diagnoser.done();

    return out;
  }

  /**Diagnoses the entire given folder
   * @param folder The folder to retrieve files of
   * @param ignores The pattern to ignore on files*/
  processFolder(folder: string, ignores: MCIgnore): boolean {
    const files = this.context.getFiles(folder, ["*"], ignores);
    let out = false;

    for (let I = 0; I < files.length; I++) {
      out = this.process(files[I]) || out;
    }

    return out;
  }

  /**Process the entire given pack
   * @param pack The pack to process
   * @returns True or false is something was processed*/
  processPack(pack: Pack): boolean {
    return this.processFolder(pack.folder, pack.context.ignores);
  }
}
