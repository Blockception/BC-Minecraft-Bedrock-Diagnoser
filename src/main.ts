export { DiagnoserContextCarrier } from "./Lib/Types/ContextCarrier";
export { DiagnoserContext, InternalDiagnosticsBuilder, Diagnoser, DiagnosticSeverity, DiagnosticsBuilder, DiagnosticsBuilderContent } from "./Lib/Types";
