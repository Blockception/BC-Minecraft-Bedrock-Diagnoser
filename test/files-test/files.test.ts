import { Documents, ProjectData } from "bc-minecraft-bedrock-project";
import { MCAttributes, MCDefinition, MCIgnore, MCProject } from "bc-minecraft-project";
import FastGlob from "fast-glob";
import { readFileSync } from "fs";
import path from "path";
import pm from "picomatch";
import { TextDocument } from "vscode-languageserver-textdocument";
import { Diagnoser, DiagnoserContext, DiagnosticsBuilderContent, ManagedDiagnosticsBuilder } from "../../src/main";
import { TestDocumentDiagnoser } from "../diagnoser";

class DocumentManager implements Documents<TextDocument> {
  getDocument(uri: string): TextDocument | undefined {
    const doc = readFileSync(uri, { encoding: "utf-8" });

    return TextDocument.create(uri, identifyDocument(uri), 1, doc);
  }
  getFiles(folder: string, patterns: string[], ignores: MCIgnore): string[] {
    return Glob.getFiles(patterns, ignores.patterns, folder);
  }
}

class TestContext implements DiagnosticsBuilderContent<TextDocument>, DiagnoserContext<TextDocument> {
  public diagnosers: TestDocumentDiagnoser<TextDocument>[] = [];

  constructor(private documentManager: DocumentManager, private projectData: ProjectData) {}
  getDocument(uri: string): TextDocument | undefined {
    return this.documentManager.getDocument(uri);
  }
  getFiles(folder: string, patterns: string[], ignores: MCIgnore): string[] {
    return this.getFiles(folder, patterns, ignores);
  }
  getCache(): ProjectData {
    return this.projectData;
  }
  getDiagnoser(doc: TextDocument, project: MCProject): ManagedDiagnosticsBuilder<TextDocument> | undefined {
    const diagnoser = new TestDocumentDiagnoser(doc, this, project);
    this.diagnosers.push(diagnoser);
    return diagnoser;
  }
}

namespace Glob {
  const opt: pm.PicomatchOptions = {
    contains: true,
  };

  /**
   *
   * @param source
   * @param ignores
   * @returns
   */
  export function excludes(source: string[], ignores: string[]): string[] {
    return source.filter((x) => !pm.isMatch(x, ignores, opt));
  }

  /**
   *
   * @param source
   * @param ignores
   * @param cwd The working directory
   * @param baseNameMatch
   * @returns
   */
  export function getFiles(
    source: string | string[],
    ignores: string[] | undefined = undefined,
    cwd: string | undefined = undefined,
    baseNameMatch: boolean | undefined = undefined
  ): string[] {
    const options: FastGlob.Options = { onlyFiles: true, absolute: true, cwd: cwd, baseNameMatch: baseNameMatch };
    let entries = FastGlob.sync(source, options);

    if (ignores && ignores.length > 0) entries = excludes(entries, ignores);

    return entries;
  }
}

namespace MinecraftFormat {
  export function getManifests(folder: string, ignores: string[]): string[] {
    return Glob.getFiles(["manifest.json", "**/manifest.json"], ignores, folder, true);
  }
  export function getBehaviorPackFiles(folder: string, ignores: string[]): string[] {
    return Glob.getFiles(
      ["**/*.{json,jsonc,json5}", "*.{json,jsonc,json5}", "*.mcfunction", "**/*.mcfunction", "**/*.lang", "*.lang"],
      ignores,
      folder
    );
  }
  export function getResourcePackFiles(folder: string, ignores: string[]): string[] {
    return Glob.getFiles(["**/*.{json,jsonc,json5}", "*.{json,jsonc,json5}", "**/*.lang", "*.lang"], ignores, folder);
  }
}

function identifyDocument(uri: string) {
  const ext = path.extname(uri);
  switch (ext) {
    case ".lang":
      return "bc-minecraft-language";
    case ".json":
      return "json";
    case ".mcfunction":
      return "bc-mcfunction";
    case ".molang":
      return "bc-minecraft-molang";
  }

  const filename = path.basename(uri);
  switch (filename) {
    case MCDefinition.filename:
    case MCIgnore.filename:
    case MCAttributes.filename:
      return "bc-minecraft-project";
  }

  return "bc-minecraft-other";
}

describe("Files test", () => {
  it("should be able to process and diagnose", async () => {
    const documentManager = new DocumentManager();
    const projectData = new ProjectData(documentManager);
    const mcproject = MCProject.createEmpty();
    const testContext = new TestContext(documentManager, projectData);
    const diagnoser = new Diagnoser(testContext);

    // Process
    const manifests = MinecraftFormat.getManifests(__dirname, mcproject.ignores.patterns);
    expect(manifests).toHaveLength(2);

    // process packs
    manifests.forEach((m) => projectData.addPack(m, mcproject));

    const bpFiles = MinecraftFormat.getBehaviorPackFiles(path.join(__dirname, "test-bp"), mcproject.ignores.patterns);
    const rpFiles = MinecraftFormat.getResourcePackFiles(path.join(__dirname, "test-rp"), mcproject.ignores.patterns);
    const files = [...bpFiles, ...rpFiles];
    files.forEach((f) => {
      console.log("processing", f);
      const t = documentManager.getDocument(f);
      expect(t).toBeDefined();
      if (!t) return;

      projectData.process(t);
    });

    // diagnose projects
    files.push(...manifests);
    files.forEach((f) => diagnoser.process(f));

    // validate
    let count = 0;
    for (const diag of testContext.diagnosers) {
      console.log("validating", diag.document.uri);
      diag.expectDone();
      count += diag.items.length;
    }

    expect(count).toBeGreaterThan(0);
  });
});
