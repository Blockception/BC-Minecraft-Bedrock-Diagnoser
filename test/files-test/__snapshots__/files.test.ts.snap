// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Files test testing file: {/test-bp/animation_controllers/controller.npc.json} expect specific errors 1`] = `[]`;

exports[`Files test testing file: {/test-bp/animations/animation.item.json} expect specific errors 1`] = `[]`;

exports[`Files test testing file: {/test-bp/entities/sample.json} expect specific errors 1`] = `
[
  {
    "code": "behaviorpack.entity.component_group.missing",
    "message": "Event is calling component group: attack_cooldown, but the component group was not found",
    "position": "events/add_missng_component_groups/attack_cooldown",
    "severity": 2,
  },
]
`;

exports[`Files test testing file: {/test-bp/functions/test.mcfunction} expect specific errors 1`] = `[]`;

exports[`Files test testing file: {/test-bp/items/apple.json} expect specific errors 1`] = `[]`;

exports[`Files test testing file: {/test-bp/manifest.json} expect specific errors 1`] = `
[
  {
    "code": "minecraft.manifest.version.minimum",
    "message": "By convention, the version numbering needs to be atleast [1, 0, 0]",
    "position": "header/version/0",
    "severity": 2,
  },
  {
    "code": "minecraft.manifest.version.minimum",
    "message": "By convention, the version numbering needs to be atleast [1, 0, 0]",
    "position": "modules/version/0",
    "severity": 2,
  },
]
`;

exports[`Files test testing file: {/test-bp/manifest.json} expect specific errors 2`] = `
[
  {
    "code": "minecraft.manifest.version.minimum",
    "message": "By convention, the version numbering needs to be atleast [1, 0, 0]",
    "position": "header/version/0",
    "severity": 2,
  },
  {
    "code": "minecraft.manifest.version.minimum",
    "message": "By convention, the version numbering needs to be atleast [1, 0, 0]",
    "position": "modules/version/0",
    "severity": 2,
  },
]
`;

exports[`Files test testing file: {/test-rp/entity/sample.entity.json} expect specific errors 1`] = `
[
  {
    "code": "resourcepack.models.missing",
    "message": "Cannot find resourcepack models definition: "geometry.sample"",
    "position": ""geometry.sample"",
    "severity": 3,
  },
  {
    "code": "behaviorpack.item.components.texture_not_found",
    "message": "Texture reference "spawn_egg_sample" was not defined in item_texture.json",
    "position": "description/spawn_egg/spawn_egg_sample",
    "severity": 3,
  },
  {
    "code": "resourcepack.texture.missing",
    "message": "Cannot find file: textures/entity/sample/sample",
    "position": "bc:sample/textures/entity/sample/sample",
    "severity": 3,
  },
]
`;

exports[`Files test testing file: {/test-rp/manifest.json} expect specific errors 1`] = `
[
  {
    "code": "minecraft.manifest.version.minimum",
    "message": "By convention, the version numbering needs to be atleast [1, 0, 0]",
    "position": "header/version/0",
    "severity": 2,
  },
  {
    "code": "minecraft.manifest.version.minimum",
    "message": "By convention, the version numbering needs to be atleast [1, 0, 0]",
    "position": "modules/version/0",
    "severity": 2,
  },
]
`;

exports[`Files test testing file: {/test-rp/manifest.json} expect specific errors 2`] = `
[
  {
    "code": "minecraft.manifest.version.minimum",
    "message": "By convention, the version numbering needs to be atleast [1, 0, 0]",
    "position": "header/version/0",
    "severity": 2,
  },
  {
    "code": "minecraft.manifest.version.minimum",
    "message": "By convention, the version numbering needs to be atleast [1, 0, 0]",
    "position": "modules/version/0",
    "severity": 2,
  },
]
`;

exports[`Files test testing file: {/test-rp/render_controllers/sample.render_controllers.json} expect specific errors 1`] = `[]`;
